### PR TITLE
[MNG-7597] Fix infinite loop when iterating PropertiesAsMap

### PR DESCRIPTION
**Pull request by gnodet**
JIRA issue: https://issues.apache.org/jira/browse/MNG-7597

### CODE REVIEW

File: maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java [hunk 1]
```diff
@@ -71,6 +71,8 @@ public Entry<String, String> next()
                         {
                             throw new NoSuchElementException();
                         }
+                        Entry<Object, Object> next = this.next;
```
Comment (slawekjaranowski) [inline maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java:74]: please use another name for local variable ... it is not clear which `next` is used

File: maven-core/src/test/java/org/apache/maven/internal/impl/PropertiesAsMapTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+public class PropertiesAsMapTest
+{
+
+    @Test
+    public void testPropertiesAsMap()
+    {
+        Properties props = new Properties();
+        props.setProperty( "foo", "bar" );
+        Map<String, String> map = new HashMap<>( new PropertiesAsMap( props ) );
```
Comment (slawekjaranowski) [inline maven-core/src/test/java/org/apache/maven/internal/impl/PropertiesAsMapTest.java:36]: please add assertions.

maybe also test for confirm, that PropertiesAsMap contains only has String as key and values

File: maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java (hunk 1 above)
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java]: Fixed

File: maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java [hunk 3]
```diff
@@ -71,6 +71,8 @@ public Entry<String, String> next()
...
+                        this.next = null;
```
[the remark this answers is not in the document]
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java]: Fixed

File: maven-core/src/test/java/org/apache/maven/internal/impl/PropertiesAsMapTest.java (hunk 2 above)
Comment (gnodet) [reply maven-core/src/test/java/org/apache/maven/internal/impl/PropertiesAsMapTest.java]: Fixed

File: maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java [hunk 4]
```diff
@@ -67,28 +67,30 @@ public boolean hasNext()
                     @Override
                     public Entry<String, String> next()
                     {
-                        if ( next == null )
+                        if ( !hasNext() )
```
Comment (kwin) [inline maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java:70]: although functionally correct I still find it weird to change the state in `hasNext()`. Why not put the logic to look for the next entry in `next` and simplify `hasNext()` with a simple check if `next == null` Also I find it very confusing that `next = null` when the end of the iterator has not yet been reached

Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/internal/impl/PropertiesAsMap.java]: Fixed

**Review by kwin — APPROVED** [review]
Thanks, looks much cleaner now.

**Comment by jira-importer** [comment]
Resolve #8858

**Pull request merged**